### PR TITLE
[RFC] Fix system('...') quoting for cmd.exe

### PR DESCRIPTION
**Pull request by equalsraf**
This is an alternative PR to #6359, it works by disabling libuv quoting only for `argv[0]=cmd.exe`. Its a workaround, but since `cmd.exe` is a special case it doesn't seem right to make `system('')` harder to use for all other shells.

In a nutshell

- Whenever spawning cmd.exe libuv argv quoting is disabled
- This affects both `system('...')` and `system([])` but only if argv[0] is `cmd.exe`

### Advantages vs #6359:

- the behaviour of `system('...')` when shell=cmd.exe should be compatible with Vim i.e. just like #6359
- current behaviour of `powershell` in `system('...')` remains as is
- user can call `system(['cmd.exe', ...])` and quoting will be disabled - not sure if this is a good or a bad thing, but at least the caller can do their own thing with regards to quoting it

### Disadvantages:

- ~~If the user sets shell to cmd.exe in some other path, then he will face the same issues we have seen so far - #6329~~ mitigated by `os_shell_is_cmdexe()`

### Other Alternatives

- parameterize this behaviour in the relevant calls. This could be possible for `jobstart()/` but `system()` still respects the original Vim api signature.
- use an option to disable process argv quoting in `system('...')/jobstart('...')` (@erw7 suggested shellxquote https://github.com/neovim/neovim/pull/6359#discussion_r108112574) but to me this sounds hard to grasp, and might make the shell options even more confusing.

### CODE REVIEW

**Comment by equalsraf** [comment]
The failling test is the same as in #6359 `system('echo a & echo b')` returns `a & echo b\n` instead of `a\nb\n`but it seems we can't get that one to work in neither Vim or Neovim so at least it is consistent. So I'm removing it.

**Comment by bounceme** [comment]
https://github.com/bounceme/windowsArgv.vim (just updated) *should* `call system(ArgvQuote(CommandLineToArgvW()))`work for anything that works in cmd.exe. it requires shellxescape and shellxquote emptied. if it does in fact work, maybe just try and see if the shell* options can be changed slightly to match

File: src/nvim/event/libuv_process.c [hunk 1]
```diff
@@ -24,6 +25,15 @@ int libuv_process_spawn(LibuvProcess *uvproc)
   if (proc->detach) {
       uvproc->uvopts.flags |= UV_PROCESS_DETACHED;
   }
+#ifdef WIN32
+  // libuv assumes spawned processes follow the convention from
+  // CommandLineToArgvW(), cmd.exe does not. Disable quoting since it will
+  // result in unexpected behaviour, the caller is left with the responsibility
+  // to quote arguments accordingly. system('') has shell* options for this.
+  if (strcmp(proc->argv[0], "cmd.exe") == 0) {
```
Comment (justinmk) [inline src/nvim/event/libuv_process.c]: could check the end of the path instead of an exact match. Also `$COMSPEC` ?

**Comment by justinmk** [comment]
This is only an improvement vs our current situation, so it's a good compromise and a great first step.

> system('echo a & echo b') returns a & echo b\n instead of a\nb\nbut it seems we can't get that one to work in neither Vim or Neovim so at least it is consistent. So I'm removing it.

Instead of removing it, make it succeed and put `-- TODO: ...` comment on it. If it fails later, we have useful information.

> user can call system(['cmd.exe', ...]) and quoting will be disabled 

But "libuv quoting" will be "enabled", right? `UV_PROCESS_WINDOWS_VERBATIM_ARGUMENTS` is _not_ specified.

> use an option to disable process argv quoting in

Let's wait until a use-case is proven before doing that. `system([])` puts the control in plugin authors' hands and makes it possible to develop a robust plugin which can be field-tested and later adopted in core if necessary.

File: runtime/doc/eval.txt [hunk 2]
```diff
@@ -4856,7 +4861,12 @@ jobstart({cmd}[, {opts}])				{Nvim} *jobstart()*
 		   :call jobstart(['ping', 'neovim.io'])
 <		If it is a path (not a name), it must include the extension: >
 		   :call jobstart(['System32\ping.exe', 'neovim.io'])
-<
+
+<		NOTE: On Windows the List {cmd} is converted according to the
+		standard command line argument format (used in CommandLineToArgvW).
+		Some programs such as cmd.exe do not, and may behave
```
Comment (brcolow) [inline runtime/doc/eval.txt]: I think this sentence is a bit confusing - some programs such as cmd.exe do not *what*?

**Comment by janlazo** [comment]
Do we have `system('cmd /c "echo a & echo b"')` for `a\nb\n`?

**Comment by bounceme** [comment]
it seems `shellxescape` just doesn't do it's job. `let &shellxescape = '' | echom system(substitute(cmd, '[()%!<>&|"^]', '^&', 'g'))` works in vim ( where this is also broken )

File: src/nvim/event/libuv_process.c (hunk 1 above)
Comment (equalsraf) [inline src/nvim/event/libuv_process.c]: pusing a commit for COMSPEC with  a test

Comment (equalsraf) [inline src/nvim/event/libuv_process.c]: > Also, STRICMP (case-insensitive) instead, right?

iiirc yes

**Comment by equalsraf** [comment]
>  > user can call system(['cmd.exe', ...]) and quoting will be disabled

> But "libuv quoting" will be "enabled", right? UV_PROCESS_WINDOWS_VERBATIM_ARGUMENTS is not specified.

No. "libuv quoting" is disabled for any situation where argv[0]=cmd.exe even if calling `system(['cmd.exe'])`. Its the only way to make a sane call to cmd.exe.

File: src/nvim/event/libuv_process.c (hunk 1 above)
Comment (equalsraf) [inline src/nvim/event/libuv_process.c]: seems like the only way to get stricmp is to include vim.h

**Comment by justinmk** [comment]
Got it. Anything left to do here or should we start with this ?

**Comment by equalsraf** [comment]
> Got it. Anything left to do here or should we start with this ?

Let me address the other comments

- [ ] stricmp
- [ ] allow for 'cmd' without and extension
- [ ] be explicit in jobstart() docs

File: src/nvim/event/libuv_process.c (hunk 1 above)
Comment (justinmk) [inline src/nvim/event/libuv_process.c]: we could move those to strings.h

**Comment by equalsraf** [comment]
All done here, the only remaining issues is lint for the original Vim code.

**Comment by justinmk** [comment]
Tested this, it fixes the known vim-plug issue, as well as other plugins I tested (gtfo). However 'shellxquote'` must be empty. Personally I'm ok with leaving it empty by default, just wanted to mention it.

**Comment by equalsraf** [comment]
> However 'shellxquote'` must be empty.

Strange, do you have some specific case I can test with?

Disabling the default shellxquote also disables shellxescape which I would assume to be needed.

**Comment by justinmk** [comment]
> Disabling the default shellxquote also disables shellxescape which I would assume to be needed.

@equalsraf I think you nailed it. vim-plug actually adds the parens explicitly, regardless of shellxquote: 
https://github.com/junegunn/vim-plug/blob/master/plug.vim#L1988

**Comment by equalsraf** [comment]
I see the issue with vim-plug too, here are some commands that were issued (I'm just doing a single PlugInstall)

```
cmd.exe /c (^(git --version^))
cmd.exe /c (^(git clone --depth 1  ^"https://git::^@github.com/junegunn/vim-easy-align.git^" ^"c:\msys64\home\dummy\neovim\build\test\plugged\vim-easy-align^" 2^>^&1^))
```

I'm not sure how they were created, for example if I call `system('git --version')` manually the spawned process is

```
cmd.exe /c (git --version)
```

and it executes successfully. So where did the second set of parenthesis come from? I found the following in vim-plug

```vim
function! s:system(cmd, ...)
  try
    let [sh, shrd] = s:chsh(1)
    let cmd = a:0 > 0 ? s:with_cd(a:cmd, a:1) : a:cmd
    return system(s:is_win ? '('.cmd.')' : cmd)
  finally
    let [&shell, &shellredir] = [sh, shrd]
  endtry
endfunction
```

i.e. it seems in windows `vim-plug` wraps the command in `()` regardless of shellxquote. Adding a check for shellxquote in that function is not sufficient to address it, I now get

```
cmd.exe /c (git --version)
cmd.exe /c (git clone --depth 1 --no-single-branch  ^"https://git::^@github.com/junegunn/vim-easy-align.git^" ^"c:\msys64\home\dummy\neovim\build\test\plugged\vim-easy-align^" 2^>^&1)
```

The git clone command runs but fails with

```
    Too many arguments.
    
    usage: git clone [<options>] [--] <repo> [<dir>]
...
```

To be clear the redirect is not our doing, it is set by vim-plug in the command passed to system. The problem at this point seems to be the use of the `&` which I was not able to fix yet.

**Comment by equalsraf** [comment]
BTW the issues with `&` exist in both Vim and Neovim (with the defaults in this PR) they just manifest a bit differently because process spawning from `system()` is different in Vim/Neovim.

**Comment by equalsraf** [comment]
Reproduceable test case, that works in Vim but not in Neovim

```vim
system('echo a 2>&1')
" Vim echoes: a
" (this PR) echoes: a 2>&1
```

In neovim this spawns

```
cmd.exe /c (echo a 2^>^&1)
```

**Comment by bounceme** [comment]
Just a thought, maybe the pipes are also unwrapping the escapes?

**Comment by bounceme** [comment]
also I don't think we need shellxquote set to anything at all. the parentheses are just for grouping multiline commands, and cause inner parens to break

**Comment by justinmk** [comment]
👍

**Comment by equalsraf** [comment]
> Just a thought, maybe the pipes are also unwrapping the escapes?

Doesn't seem to be the case, I've changed the default shellxescape and its passed all the tests even the ones that I previously marked as broken.

I suspect I have been looking at this wrong, and since we don't run the multiple instances of cmd like vim,  we want the special characters to be left unescaped (with `^`) in order to have the intended effect (redirection and so on). But in the morning I migh have a clearer mind on this.

> also I don't think we need shellxquote set to anything at all. the parentheses are just for grouping multiline commands

This is my one remaining doubt - do we need to wrap the command in something (`()` or `""`) and does nesting these tokens cause any problem?

> and cause inner parens to break

 Do you have an test case? With `shellxquote=(` and `shellxescape=` I was able to run

- `system('(echo a) & echo b')` -> `cmd.exe /c ((echo a) & echo b)` and prints `a\nb\n`
- removing one of the inner parens causes an error, which is what I would expect from the shell

Similarly with `shellxquote="` I was able to run commands with nested quotes

- `system('echo "a b"')` -> `cmd.exe /c "echo "a b""` and prints `"a b"`
- unbalanced quote `system('echo "a b')` -> `cmd.exe /c "echo "a b"` and prints `"a b`

that last one would fail if shellxquote was `(`.

Pushing my last commit for the nigh, changing the default for shellxquote to "".

**Comment by bounceme** [comment]
`echo (a)` vs`(echo (a))` in the actual cmd.exe, I think the `&` `((echo a) & echo b)` separates the two commands, as an argument it gives an error

**Comment by justinmk** [comment]
Merged in #6497. Thank you @equalsraf, this is a simpler, more elegant solution than I would have hoped for.

**Pull request closed without merging**